File: packages/backend/src/MeanList.ts
import { Maybe, Types, timestamp } from '@dotstats/common';

export default class MeanList<T extends number> {
  private periodCount = 0;
  private periodSum = 0;
  private meanIndex = 0;
  private means = Array<T>(20).fill(0 as T);
  private ticksPerMean = 1;

  /**
   * Push a new value, returns true if a new mean value was produced
   *
   * @param  {T}       value
   *
   * @return {boolean}
   */
  public push(val: Maybe<T>): boolean {
    if (val == null) {
      return false;
    } 

    if (this.meanIndex === 20 && this.ticksPerMean < 32) {
      this.squashMeans();
    }

    this.periodSum += val as number;
    this.periodCount += 1;

    if (this.periodCount === this.ticksPerMean) {
      this.pushMean();
      return true;
    }

    return false;
  }

  public get(): Array<T> {
    if (this.meanIndex === 20) {
      return this.means;
    } else {
      return this.means.slice(0, this.meanIndex);
    }
  }

  private pushMean() {
    const mean = (this.periodSum / this.periodCount) as T;

    if (this.meanIndex === 20 && this.ticksPerMean === 32) {
      this.means.copyWithin(0, 1);
      this.means[19] = mean;
    } else {
      this.means[this.meanIndex++] = mean;
    }

    this.periodSum = 0;
    this.periodCount = 0;
  }

  private squashMeans() {
    this.ticksPerMean *= 2;

    const means = this.means;

    for (let i = 0; i < 10; i++) {
      let i2 = i * 2;
      means[i] = (((means[i2] as number) + (means[i2 + 1] as number)) / 2) as T;
    }

    this.meanIndex = 10;
  }
}

File: backend/src/util/mean_list.rs
use num_traits::{Float, Zero};
use std::ops::AddAssign;

pub struct MeanList<T>
where
    T: Float + AddAssign + Zero + From<u8>,
{
    period_sum: T,
    period_count: u8,
    mean_index: u8,
    means: [T; 20],
    ticks_per_mean: u8,
}

impl<T> Default for MeanList<T>
where
    T: Float + AddAssign + Zero + From<u8>,
{
    fn default() -> MeanList<T> {
        MeanList {
            period_sum: T::zero(),
            period_count: 0,
            mean_index: 0,
            means: [T::zero(); 20],
            ticks_per_mean: 1,
        }
    }
}

impl<T> MeanList<T>
where
    T: Float + AddAssign + Zero + From<u8>,
{
    pub fn slice(&self) -> &[T] {
        &self.means[..usize::from(self.mean_index)]
    }

    pub fn push(&mut self, val: T) -> bool {
        if self.mean_index == 20 && self.ticks_per_mean < 32 {
            self.squash_means();
        }

        self.period_sum += val;
        self.period_count += 1;

        if self.period_count == self.ticks_per_mean {
            self.push_mean();
            true
        } else {
            false
        }
    }

    fn push_mean(&mut self) {
        let mean = self.period_sum / std::convert::From::from(self.period_count);

        if self.mean_index == 20 && self.ticks_per_mean == 32 {
            self.means.rotate_left(1);
            self.means[19] = mean;
        } else {
            self.means[usize::from(self.mean_index)] = mean;
            self.mean_index += 1;
        }

        self.period_sum = T::zero();
        self.period_count = 0;
    }

    fn squash_means(&mut self) {
        self.ticks_per_mean *= 2;
        self.mean_index = 10;

        for i in 0..10 {
            let i2 = i * 2;

            self.means[i] = (self.means[i2] + self.means[i2 + 1]) / std::convert::From::from(2)
        }
    }
}
